lib/semantics/tools.cc

```diff
@@ -382,4 +354,26 @@ const Symbol *HasEventOrLockPotentialComponent(
   }
   return nullptr;
 }
+
+const Symbol *FindUltimateComponent(const DerivedTypeSpec &derivedTypeSpec,
```

Comment:
Alternatively, each `DerivedTypeSpec` could have a list of ultimate components, and you could use `find_if()` on it.  (But I think that this traversal with a predicate is better, just wanted to raise the possibility in case you like it.)